README.md

```diff
@@ -1 +1,142 @@
+    toppings = fields.String(required=False)
+```
+
+Instead of marshmallow, FastAPI makes extensive use of `pydantic` to provide the validation layers. Pydantic is a more modern library in comparison. It uses python typehints in order to define expected field types and has more built-in functionality when compared to marshmallow. It's straightforward to convert the above definitions into ones that are pydantic compatible. Note that we remove the `PizzaSearchSchema` entirely because this definition will be specified in another file.
+
+```
+from pydantic import BaseModel
+from uuid import UUID
+
+class NewPizza(BaseModel):
+    toppings: str
+
+class Pizza(NewPizza)
+    id: UUID
+```
+
+Just like typehinting in standard python functions, arguments are required unless you specify an `Optional` flag alongside their type. This will enforce that client callers provide `toppings` when creating a new `Pizza`.
+
+### Routes
+
+In microcosm-flask, your routes are usually split between two files. You'll have `pizza/crud.py` and `pizza/controller.py`. The crud file specifies the supported operations and resources for the given namespace. The controller will implement any relevant business logic to transform the input client request before passing it to the backing store. Something like:
+
+```
+@binding("pizza_v1_routes")
+def configure_pizza_routes(graph):
+    controller = graph.credential_pack_controller
+
+    mappings = {
+        Operation.Create: EndpointDefinition(
+            func=transactional(controller.create),
+            request_schema=NewPizzaSchema(),
+            response_schema=PizzaSchema(),
+        ),
+        Operation.Retrieve: EndpointDefinition(
+            func=controller.retrieve, 
+            response_schema=PizzaSchema(),
+        ),
+        Operation.Search: EndpointDefinition(
+            func=controller.search,
+            request_schema=SearchPizzaSchema(),
+            response_schema=PizzaSchema(),
+        ),
+    }
+    configure_crud(graph, controller.ns, mappings)
+
+    return controller.ns
+```
+
+```
+@binding("pizza_controller")
+class PizzaController(CRUDStoreAdapter):
+    def __init__(self, graph):
+        super().__init__(graph, graph.pizza_store)
+        self.ns = Namespace(subject=Pizza, version="v1")
+```
+
+One drawback with this approach is that a lot of the logic is abstracted away into the `CRUDStoreAdapter` and `configure_crud` code. It's not immediately transparent to new team members what the API functions will actually look like when they're created.
+
+The goal in our new routing convention is to have one file the provides the full source of truth. This route will contain an explicit definition of all APIs that are available for the given database object. The typehinting of both the function and the response signatures are parsed by `microcosm-fastapi` for you - requests are validated against the function types and responses are serialized to fit within the return type annotation.
```

Comment:
I'm not convinced by the need to have those explicit definitions; if it ends up being the same boilerplate calling the `super()` methods in most cases I don't see what we gain there. I tend to think most of the complexity/magic of the routing is hidden away in `configure_crud`, which we still have here.

Reply:
@afallou - I agree in principle when we're mainly just calling `super()`. In practice we need these definitions because it provides the type hinting necessary for FastAPI to serialize the input/return data.

In theory we could refactor some of that logic into `configure_crud` and get even more magical about it - but here I think this design has a few nice upshots:

- You only have to look in one file for a given namespace's API contract
- `configure_crud` stays simple, there's none of this magic logic to get the input/output schemas and serialize to them